src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,257 @@
+        int r;
+
+        /* If NULL is passed we query the default solution, which is the first one listed
+         * in the SOLUTIONS key.
+         */
+        if (name == NULL) {
+                r = get_known_solutions(device, &solutions);
+
+                /* A positive 1 return indicates no solutions exist for the device, which should not
+                 * be treated as an error.
+                 */
+                if (r < 0)
+                        return r;
+                else if (r == 1)
+                        return r;
+
+                name = ASSERT_PTR(solutions[0]);
+        }
+
+        upper_name = strdup(name);
+        if (!upper_name)
+                return log_oom();
+
+        ascii_strupper(upper_name);
+
+        if (asprintf(&qos_key, "IOCOST_QOS_%s", upper_name) < 0)
+                return log_oom();
+
+        if (asprintf(&model_key, "IOCOST_MODEL_%s", upper_name) < 0)
+                return log_oom();
+
+        r = sd_device_get_property_value(device, model_key, model);
+        if (r < 0) {
+                log_error("Model key missing from hwdb for iocost solution '%s'.", name);
+                return -1;
+        }
+
+        r = sd_device_get_property_value(device, qos_key, qos);
+        if (r < 0) {
+                log_error("QoS key missing from hwdb for iocost solution '%s'.", name);
+                return -1;
+        }
+
+        return 0;
+}
+
+static int apply_solution_for_path(const char *path, const char *name) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *model = NULL, *qos = NULL;
+        const char *model_params = NULL, *qos_params = NULL;
+        dev_t devnum;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = query_named_solution(device, name, &model_params, &qos_params);
+        if (r < 0)
+                return r;
+        else if (r == 1) {
+                log_info("Parameters for iocost model missing from hwdb, doing nothing.");
+                return 0;
+        }
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        if (asprintf(&model, DEVNUM_FORMAT_STR " model=linear ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), model_params) < 0)
+                return log_oom();
+
+        if (asprintf(&qos, DEVNUM_FORMAT_STR " enable=1 ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), qos_params) < 0)
+                return log_oom();
+
+        log_debug("Applying iocost parameters to %s using solution '%s'\n\tio.cost.model: %s\n\tio.cost.qos: %s\n", path, name ? name : "default", model, qos);
+
+        r = cg_set_attribute("io", NULL, "io.cost.qos", qos);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set qos: %m");
```

Comment:
these should probably be handled gracefully. i.e. if we get ENOENT then this means the io controller might not be available, but that's entirely OK. hence maybe:

```c
log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_ERROR, r, …`
return r == -ENOENT ? 0 : r
```
i.e. make this a debug message in case of ENOENT, and otherwise log an error